# TASK 2 : Simple Text Classification

In [1]:
def gender_features(word):
    return {'last_letter' : word[-1]}

In [2]:
gender_features('Avina')

{'last_letter': 'a'}

In [3]:
gender_features('Obama')

{'last_letter': 'a'}

In [ ]:
gender_features('Obama')

{'last_letter': 'a'}

In [8]:
import nltk
nltk.download('names')

[nltk_data] Downloading package names to /root/nltk_data...
[nltk_data]   Unzipping corpora/names.zip.


True

In [9]:
len(names.words())

7944

In [11]:
labelled_names = ([(name,'male') for name in names.words('male.txt')] + [(name, 'female')  for name in names.words('female.txt')])

In [12]:
labelled_names

[('Aamir', 'male'),
 ('Aaron', 'male'),
 ('Abbey', 'male'),
 ('Abbie', 'male'),
 ('Abbot', 'male'),
 ('Abbott', 'male'),
 ('Abby', 'male'),
 ('Abdel', 'male'),
 ('Abdul', 'male'),
 ('Abdulkarim', 'male'),
 ('Abdullah', 'male'),
 ('Abe', 'male'),
 ('Abel', 'male'),
 ('Abelard', 'male'),
 ('Abner', 'male'),
 ('Abraham', 'male'),
 ('Abram', 'male'),
 ('Ace', 'male'),
 ('Adair', 'male'),
 ('Adam', 'male'),
 ('Adams', 'male'),
 ('Addie', 'male'),
 ('Adger', 'male'),
 ('Aditya', 'male'),
 ('Adlai', 'male'),
 ('Adnan', 'male'),
 ('Adolf', 'male'),
 ('Adolfo', 'male'),
 ('Adolph', 'male'),
 ('Adolphe', 'male'),
 ('Adolpho', 'male'),
 ('Adolphus', 'male'),
 ('Adrian', 'male'),
 ('Adrick', 'male'),
 ('Adrien', 'male'),
 ('Agamemnon', 'male'),
 ('Aguinaldo', 'male'),
 ('Aguste', 'male'),
 ('Agustin', 'male'),
 ('Aharon', 'male'),
 ('Ahmad', 'male'),
 ('Ahmed', 'male'),
 ('Ahmet', 'male'),
 ('Ajai', 'male'),
 ('Ajay', 'male'),
 ('Al', 'male'),
 ('Alaa', 'male'),
 ('Alain', 'male'),
 ('Alan', 'male

In [13]:
import random
random.shuffle(labeled_names)
featuresets = [(gender_features(n), gender) for (n, gender) in labeled_names]
featuresets

[({'last_letter': 'l'}, 'female'),
 ({'last_letter': 'g'}, 'male'),
 ({'last_letter': 'e'}, 'female'),
 ({'last_letter': 'a'}, 'female'),
 ({'last_letter': 'a'}, 'female'),
 ({'last_letter': 'h'}, 'female'),
 ({'last_letter': 'y'}, 'female'),
 ({'last_letter': 'e'}, 'female'),
 ({'last_letter': 'e'}, 'female'),
 ({'last_letter': 'o'}, 'female'),
 ({'last_letter': 'y'}, 'female'),
 ({'last_letter': 'a'}, 'female'),
 ({'last_letter': 'n'}, 'male'),
 ({'last_letter': 'a'}, 'female'),
 ({'last_letter': 'n'}, 'male'),
 ({'last_letter': 'a'}, 'female'),
 ({'last_letter': 'k'}, 'male'),
 ({'last_letter': 'i'}, 'female'),
 ({'last_letter': 'h'}, 'female'),
 ({'last_letter': 'a'}, 'female'),
 ({'last_letter': 'n'}, 'male'),
 ({'last_letter': 'x'}, 'male'),
 ({'last_letter': 'd'}, 'male'),
 ({'last_letter': 'e'}, 'female'),
 ({'last_letter': 'e'}, 'male'),
 ({'last_letter': 'o'}, 'male'),
 ({'last_letter': 'r'}, 'female'),
 ({'last_letter': 'h'}, 'male'),
 ({'last_letter': 'y'}, 'female'),
 ({'l

In [14]:
train_set, test_set = featuresets[:5000], featuresets[-2000:]
import nltk
classifier = nltk.NaiveBayesClassifier.train(train_set)

In [15]:
classifier.classify(gender_features('David'))

'male'

In [16]:
classifier.classify(gender_features('Avina'))

'female'

In [18]:
classifier.classify(gender_features('Farhan'))

'male'

In [19]:
classifier.classify(gender_features('Devadutta'))

'female'

In [20]:
classifier.classify(gender_features('Ikreet'))

'male'

In [21]:
print(nltk.classify.accuracy(classifier, test_set))

0.7685


# Task 3 - Count Vectorizer

In [22]:
from sklearn.feature_extraction.text import CountVectorizer
vect = CountVectorizer(binary=True)
corpus = ["Tessaract is good optical character recognition engine  ", "optical character recognition is significant "]

In [23]:
vect.fit(corpus)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [24]:
vocab = vect.vocabulary_
vocab

{'character': 0,
 'engine': 1,
 'good': 2,
 'is': 3,
 'optical': 4,
 'recognition': 5,
 'significant': 6,
 'tessaract': 7}

In [25]:
for key in sorted(vocab.keys()):
    print("{}:{}".format(key, vocab[key]))

character:0
engine:1
good:2
is:3
optical:4
recognition:5
significant:6
tessaract:7


In [26]:
print(vect.transform(["This is a good optical illusion"]).toarray())

[[0 0 1 1 1 0 0 0]]


## Finding Similarity between documents

In [27]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vect.transform(["Google Cloud Vision is a character recognition engine"]).toarray(), vect.transform(["OCR is an optical character recognition engine"]).toarray())
print(similarity)

[[0.89442719]]


## Magnum Opus Task

In [28]:
doc1 = """Chocolate is prepared from the fruit of the Theobroma cacao, a tropical tree whose name means "food of the gods" in Greek, according to "Chocolate: Food of the Gods," an online exhibit by the Cornell University Library.
Theobroma cacao trees are native to the Amazon and Orinoco river basins in South America. The trees are widely distributed from southeastern Mexico to the Amazon River. They thrive in hot, humid areas within about 20 degrees of the equator, according to Cornell. As the popularity of chocolate spread, growers established plantations in other regions, such as West Africa and South and Southeast Asia. Today, Ghana, Côte d'Ivoire, Nigeria, Indonesia and Brazil account for 79 percent of the world's cacao production.
Cacao trees bear fruit that are about the same size and shape of a papaya, according to Patric Chocolate. These bumpy, lumpy berries, or pods, are full of up to 50 sour seeds, or beans, covered in white pulp."""

In [35]:
doc2="""But no matter the origin of the cocoa beans, whether it’s a blend of Latin American and African cacao, or a single-origin chocolate bar, the process for making chocolate is essentially the same. And it starts with the tree, pod, and bean.Chocolate begins with the cacao tree (Theobroma cacao), which grows within 20° north and south of the Equator and thrives on a mix of hot temperatures, rain, and shade.
Each tree bears oval fruits, or pods, which are about 5–12 inches long. Each pod contains 30–50 seeds, and it’s these seeds the world knows as cacao (or cocoa) beans.
What is the difference between cocoa and cacao? The tree, pod, and bean/seed are typically referred to as “cacao,” while the word “cocoa” is reserved for the bean after it has been fermented, dried, and roasted"""

### With Stemming

In [30]:
from nltk.stem import LancasterStemmer
ls = LancasterStemmer()

In [55]:
d1 = [ls.stem(w) for w in doc1.split()]
d1 = " ".join(d1)
d2 = [ls.stem(w) for w in doc2.split()]
d2 = " ".join(d2) 

In [56]:
doc = [d1,d2]

In [38]:
doc

['chocol is prep from the fruit of the theobrom cacao, a trop tre whos nam mean "food of the gods" in greek, accord to "chocolate: food of the gods," an onlin exhibit by the cornel univers library. theobrom cacao tre ar nat to the amazon and orinoco riv basin in sou america. the tre ar wid distribut from southeastern mexico to the amazon river. they thrive in hot, humid area within about 20 degr of the equator, accord to cornell. as the popul of chocol spread, grow est plant in oth regions, such as west afric and sou and southeast asia. today, ghana, côte d\'ivoire, nigeria, indones and brazil account for 79 perc of the world\'s cacao production. cacao tre bear fruit that ar about the sam siz and shap of a papaya, accord to pat chocolate. thes bumpy, lumpy berries, or pods, ar ful of up to 50 sour seeds, or beans, cov in whit pulp.',
 'but no mat the origin of the coco beans, wheth it’s a blend of latin am and afr cacao, or a single-origin chocol bar, the process for mak chocol is ess 

In [62]:
from sklearn.feature_extraction.text import CountVectorizer
import pandas as pd
vect=CountVectorizer(binary=True)

In [60]:
vect.fit(doc)

CountVectorizer(analyzer='word', binary=True, decode_error='strict',
                dtype=<class 'numpy.int64'>, encoding='utf-8', input='content',
                lowercase=True, max_df=1.0, max_features=None, min_df=1,
                ngram_range=(1, 1), preprocessor=None, stop_words=None,
                strip_accents=None, token_pattern='(?u)\\b\\w\\w+\\b',
                tokenizer=None, vocabulary=None)

In [61]:
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vect.transform([d1]).toarray(), vect.transform([d2]).toarray())
similarity

array([[0.28061004]])

### Without Stemming

In [63]:
doc = [doc1,doc2]
doc

['Chocolate is prepared from the fruit of the Theobroma cacao, a tropical tree whose name means "food of the gods" in Greek, according to "Chocolate: Food of the Gods," an online exhibit by the Cornell University Library.\nTheobroma cacao trees are native to the Amazon and Orinoco river basins in South America. The trees are widely distributed from southeastern Mexico to the Amazon River. They thrive in hot, humid areas within about 20 degrees of the equator, according to Cornell. As the popularity of chocolate spread, growers established plantations in other regions, such as West Africa and South and Southeast Asia. Today, Ghana, Côte d\'Ivoire, Nigeria, Indonesia and Brazil account for 79 percent of the world\'s cacao production.\nCacao trees bear fruit that are about the same size and shape of a papaya, according to Patric Chocolate. These bumpy, lumpy berries, or pods, are full of up to 50 sour seeds, or beans, covered in white pulp.',
 'But no matter the origin of the cocoa beans,

In [64]:
vect.fit(doc)
from sklearn.metrics.pairwise import cosine_similarity
similarity = cosine_similarity(vect.transform([doc1]).toarray(), vect.transform([doc2]).toarray())
similarity

array([[0.29034426]])

The similarity between the documents with stemming is 28.06% and without stemming is 29.03%. Thus we can conclude that stemming decreases the similarity percentage in the documents.